File: backend/app/models/aliquot.py
"""
Aliquot Model - All fields use camelCase
"""
from sqlalchemy import Column, String, Float, DateTime, JSON, Enum, ForeignKey
from sqlalchemy.sql import func
from app.database import Base
from app.schemas.enums import AliquotStatus, ContainerType


class Aliquot(Base):
    __tablename__ = "aliquots"

    aliquotId = Column("aliquot_id", String, primary_key=True, index=True)  # ALQ-YYYYMMDD-XXX
    parentSampleId = Column("parent_sample_id", String, ForeignKey("samples.sample_id"), nullable=False, index=True)
    orderId = Column("order_id", String, ForeignKey("orders.order_id"), nullable=False, index=True)
    patientId = Column("patient_id", String, ForeignKey("patients.id"), nullable=False)

    # Aliquot details
    aliquotNumber = Column("aliquot_number", Float, nullable=False)  # 1, 2, 3, etc.
    volume = Column(Float, nullable=False)  # mL in this aliquot
    remainingVolume = Column("remaining_volume", Float, nullable=False)  # mL remaining after testing

    # Purpose
    linkedTestCodes = Column("linked_test_codes", JSON, nullable=False)  # Array of test codes
    purpose = Column(String, nullable=True)  # e.g., "Sendout tests", "Chemistry panel"

    # Container
    containerType = Column("container_type", Enum(ContainerType), nullable=False)
    barcode = Column(String, nullable=False, unique=True, index=True)

    # Status and location
    status = Column(Enum(AliquotStatus), nullable=False, default=AliquotStatus.AVAILABLE)
    currentLocation = Column("current_location", String, nullable=False)

    # Usage tracking
    usedForTests = Column("used_for_tests", JSON, nullable=True)  # Array of test codes that have consumed this
    consumedAt = Column("consumed_at", DateTime(timezone=True), nullable=True)
    consumedBy = Column("consumed_by", String, nullable=True)

    # Storage
    storageLocation = Column("storage_location", String, nullable=True)
    storageConditions = Column("storage_conditions", String, nullable=True)

    # Disposal
    disposedAt = Column("disposed_at", DateTime(timezone=True), nullable=True)
    disposedBy = Column("disposed_by", String, nullable=True)

    # Metadata
    createdAt = Column("created_at", DateTime(timezone=True), server_default=func.now())
    createdBy = Column("created_by", String, nullable=False)

File: frontend/src/types/aliquot.ts
// Aliquot Types - Portions of a sample for different tests/workstations

import type { ContainerType } from './sample';

export type AliquotStatus =
  | 'available'      // Ready for use
  | 'in-use'         // Currently being tested
  | 'consumed'       // Fully used up
  | 'stored'         // Archived for potential retest
  | 'disposed';      // Properly disposed

export interface Aliquot {
  aliquotId: number;               // Integer ID, displayed as ALQ{id}
  parentSampleId: number;          // Links to parent sample
  orderId: number;
  patientId: number;

  // Aliquot details
  aliquotNumber: number;           // 1, 2, 3, etc. (Aliquot 1 of 3)
  volume: number;                  // mL in this aliquot
  remainingVolume: number;         // mL remaining after testing

  // Purpose
  linkedTestCodes: string[];       // Tests assigned to this aliquot
  purpose?: string;                // e.g., "Sendout tests", "Chemistry panel"

  // Container
  containerType: ContainerType;
  barcode: string;

  // Status and location
  status: AliquotStatus;
  currentLocation: string;

  // Creation tracking
  createdAt: string;
  createdBy: string;

  // Usage tracking
  usedForTests: string[];          // Tests that have consumed this aliquot
  consumedAt: string | null;
  consumedBy: string | null;

  // Storage
  storageLocation?: string;
  storageConditions?: string;

  // Disposal
  disposedAt: string | null;
  disposedBy: string | null;
}

// Aliquoting plan before creating aliquots
export interface AliquotPlan {
  sampleId: number;
  totalVolume: number;
  aliquots: {
    aliquotNumber: number;
    volume: number;
    testCodes: string[];
    destination: string;           // Workstation or sendout lab
    containerType: ContainerType;
  }[];
  deadVolume: number;              // Volume lost in transfer
  totalAliquotVolume: number;      // Sum of all aliquots + dead volume
  isFeasible: boolean;             // Can we create these aliquots?
  warnings: string[];
}
